### ISSUE TITLE
Why is there no default static pool manager?

### ISSUE THREAD

**Issue by kkm000** [Microsoft/Microsoft.IO.RecyclableMemoryStream#38]
I do not know if this is just a silly question or a feature request. Suppose there are multiple separately maintained assemblies allocating `RecyclableMemoryStream`s. Should each assembly declare its own static `RecyclableMemoryStreamManager`? This goes somewhat against the grain of the idea of pooling, although I understand that block size requirements might be different. But still I assume that the defaults may be quite sane for a sizable range of applications. 

So I wonder why is there no default, per-appdomain `RecyclableMemoryStreamManager`? It would be accessed through a static property (`RecyclableMemoryStreamManager.Default`), and the `RecyclableMemoryStream` could then have a parameterless constructor (and other managerless constructors) that would use the default manager. As many other things in the BCL, the default pool may be made configurable through app.config as well.

**Comment by doubleyewdee** [comment]
Great question! The original design of this library was for an application where the use of multiple AppDomains was expressly forbidden for a variety of reasons. As such the library itself intentionally makes no effort to handle AppDomains, nor does it make recommendations for users in those scenarios. Additionally the original use-case envisioned a supervising hosting layer which would handle any requests for a MemoryStream (or underlying byte[] object) transparently to the consumer, so the desire for a default was not envisioned.

My off-the-cuff opinion is that adding a late-instantiated global (well, per-AppDomain) default manager would not be unreasonable. However, I want to balance this against the original intent for explicit use and management of the pools, So I'd like to understand a bit more: where would having a default pool be of assistance? What problems would a default pool resolve? How does it simplify (e.g.) library design or development?

My inclination is to maintain the position that pool creation and use should be explicit by the developer and coded-in with their libraries, but I certainly don't want to exclude useful scenarios we have not considered.

Thanks!

**Comment by doubleyewdee** [comment]
One more thing I'd add: there's really no circumstance I can envision where you'd want to share a pool across AppDomains. Doing so introduces substantial complexity and, critically, performance concerns which largely undermine the value added by pooling in the first place.

**Comment by IanYates** [comment]
@doubleyewdee - I think you've seen the word `AppDomain` and focussed too much on that (I did too on first reading).  @kkm000 was just noting that a static would only work within a single `AppDomain` and, as you mentioned, sharing between them wouldn't work all that well.  

I *think* his question is more along of the lines of...  Assume we have well-known library A use `RecyclableMemoryStream` and well-known library B also using `RecyclableMemoryStream`.  At the moment each library would have its own "private" `RecyclableMemoryStreamManager` instance with no opportunity to have one common pool.
This is not unlike several libraries already doing things like IL-merging their own copies of JSON.Net, but at least in that case there's no real harm apart from on-disk bloat.
It'd be like every library having to maintain their own `TaskPool` rather than using the common `TaskPool` from the BCL.  @kkm000 is asking to see if we can have a common recyclable memory stream manager, with the obvious proviso that it wouldn't work across appdomains and that, if a library really wanted their own, they can still do so by using different constructor overloads.

**Comment by stakx** [comment]
@kkm000, having a static (`AppDomain`-wide) `RecyclableMemoryStreamManager.Default` instance might be nice and indeed enable some memory usage optimizations when an application integrates several libraries that use recyclable memory streams... but I'm not sure that having the option of configuring it in `app.config` would be a good idea.

If that were possible, any library making use of the "global" manager couldn't make any assumptions about the manager it's getting. It would *have to* first check whether that manager's block size settings are appropriate for the library's purposes and, if not, still instantiate their own manager with more suitable settings. That's not a particularly nice usage pattern IMHO, even though as an "opt-in" feature it might be reasonable.

Building on that critique, it might be better to not expose the global manager, but instead a global static method that needs to be given e.g. minimal buffer sizes (or a range of suitable buffer sizes) as required by an application/library; the method might then return the global manager, or a newly instantiated one, depending on whether the global manager meets the stated requirements. Bit this is perhaps all getting a bit too clever and specific to be useful in a broad range of use cases (IMHO).

**Comment by kkm000** [comment]
@IanYates -- you are entirely correct, I mean just a static default. Your `TaskPool` analogy is correct, thanks! @doubleyewdee, sorry about the confusion. What I meant is just a static property like
```c#
class RecyclableMemoryStreamManager {
 public static RecyclableMemoryStreamManager Default => ....
...
```

@stakx -- sure, a valid point. Maybe even go simpler, and not allow configuring the default manager at all, such that with the default manager you always get the default buffer thresholds, otherwise register your own?

**Comment by kkm000** [comment]
I am changing the title to avoid the confusion I introduced with it, my bad!

**Comment by doubleyewdee** [comment]
Circling back here, sorry -- this isn't a design goal of the library and not something we intend to change at this time. As noted above there are concerns around configurability that make this impractical, In my experience using this library I have almost always found that I needed to tune it in some fashion to get optimal usage, so the value of having a global "default" seems dubious at best considering configuration changes cannot currently be made after the manager is constructed.

Closing this for now.
